### PR TITLE
Add translations from rustcc/writing-an-os-in-rust

### PR DESCRIPTION
**Pull request by phil-opp**
https://github.com/rustcc/writing-an-os-in-rust/tree/d5d1e6abc85a5a5be4be21a555f488835c1fe6d7

cc @luojia65 @Rustin-Liu

### CODE REVIEW

File: blog/content/second-edition/posts/04-testing/index.zh-CN.md [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+
+这里的问题在于，`cargo test`会将所有非`0`的错误码都视为测试失败。
+
+### 成功退出(Exit)代码
+
+为了解决这个问题， `bootimage`提供了一个 `test-success-exit-code`配置项，可以将指定的退出代码映射到退出代码`0`:
+
+```toml
+[package.metadata.bootimage]
+test-args = […]
```
Comment (bjorn3) [inline blog/content/second-edition/posts/04-testing/index.zh-CN.md]: Gh markdown syntax highlighter doesn't like the `…`

Comment (phil-opp) [reply blog/content/second-edition/posts/04-testing/index.zh-CN.md]: Not sure what we can do about that. It seems like it does not like `...` (three dots) either…

Comment (bjorn3) [reply blog/content/second-edition/posts/04-testing/index.zh-CN.md]: :( If Zola renders it fine, I guess we can live with it.

Comment (luojia65) [reply blog/content/second-edition/posts/04-testing/index.zh-CN.md]: Does this renders fine in English-version article?

Comment (phil-opp) [reply blog/content/second-edition/posts/04-testing/index.zh-CN.md]: Seems so: https://os.phil-opp.com/testing/#success-exit-code

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-title = "独立的Rust二进制"
+title = "独立式可执行程序"
 weight = 1
 path = "zh-CN/freestanding-rust-binary"
 date = 2018-02-10
 
 +++
 
-创建我们自己的操作系统内核的第一步是创建一个不链接标准库的Rust可执行文件。 这使得无需基础操作系统即可在[裸机][bare metal]上运行Rust代码。
+创建一个不连接标准库的Rust可执行文件，将是我们迈出的第一步。无需底层操作系统的支撑，这将能让在**裸机**（[bare metal]）上运行Rust代码成为现实。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]: 连接改为链接。

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 3]
```diff
@@ -20,41 +20,27 @@ date = 2018-02-10
 
 <!-- toc -->
 
-## 介绍
-要编写操作系统内核，我们需要不依赖于任何操作系统功能的代码。 这意味着我们不能使用线程，文件，堆内存，网络，随机数，标准输出或任何其他需要操作系统抽象或特定硬件的功能。这很有意义，因为我们正在尝试编写自己的OS和我们的驱动程序。
+## 简介
 
-这意味着我们不能使用大多数[Rust标准库][Rust standard library]，但是我们可以使用很多Rust功能。 例如，我们可以使用[迭代器][iterators]，[闭包][closures]，[模式匹配][pattern matching]，[option]和[result]，[string formatting]，当然也可以使用[所有权系统][ownership system]。 这些功能使以一种非常有表现力的高级方式编写内核成为可能，而无需担心[不确定的行为][undefined behavior]或[内存安全性][memory safety]。
+要编写一个操作系统内核，我们需要不基于任何操作系统特性的代码。这意味着我们不能使用线程、文件、堆内存、网络、随机数、标准输出，或其它任何需要操作系统抽象和特定硬件的特性；这其实讲得通，因为我们正在编写自己的操作系统和硬件驱动。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
要编写一个操作系统内核，我们需要编写不依赖任何操作系统特性的代码。这意味着我们不能使用线程、文件、堆内存、网络、随机数、标准输出，或其它任何需要操作系统抽象和特定硬件的特性；这其实讲得通，因为我们正在编写自己的操作系统和硬件驱动。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 4]
```diff
@@ -20,41 +20,27 @@ date = 2018-02-10
...
-[memory safety]: https://tonyarcieri.com/it-s-time-for-a-memory-safety-intervention
+实现这一点，意味着我们不能使用[Rust标准库](https://doc.rust-lang.org/std/)的大部分；但还有很多Rust特性是我们依然可以使用的。比如说，我们可以使用[迭代器](https://doc.rust-lang.org/book/ch13-02-iterators.html)、[闭包](https://doc.rust-lang.org/book/ch13-01-closures.html)、[模式匹配](https://doc.rust-lang.org/book/ch06-00-enums.html)、[Option](https://doc.rust-lang.org/core/option/)、[Result](https://doc.rust-lang.org/core/result/index.html)、[字符串格式化](https://doc.rust-lang.org/core/macro.write.html)，当然还有[所有权系统](https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html)。这些功能让我们能够编写表达性强、高层抽象的操作系统，而无需操心[未定义行为](https://www.nayuki.io/page/undefined-behavior-in-c-and-cplusplus-programs)和[内存安全](https://tonyarcieri.com/it-s-time-for-a-memory-safety-intervention)。
 
-为了在Rust中创建OS内核，我们需要创建一个无需底层操作系统即可运行的可执行文件。 此类可执行文件通常称为“独立式”或“裸机”可执行文件。
+为了用Rust编写一个操作系统内核，我们需要创建一个独立于操作系统的可执行程序。这样的可执行程序常被称作**独立式可执行程序**（freestanding executable）或**裸机程序**(bare-metal executable)。
 
-这篇文章描述了创建一个独立的Rust二进制文件的必要步骤，并解释了为什么需要这些步骤。 如果您仅对一个最小的示例感兴趣，可以 **[跳转到摘要](＃summary)**。
+在这篇文章里，我们将逐步地创建一个独立式可执行程序，并且详细解释为什么每个步骤都是必须的。如果读者只对最终的代码感兴趣，可以跳转到本篇文章的小结部分。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
在这篇文章里，我们将逐步地创建一个独立式可执行程序，并且详细解释为什么需要这样做。如果读者只对最终的代码感兴趣，可以跳转到本篇文章的小结部分。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 5]
```diff
@@ -80,22 +65,19 @@ fn main() {
...
   |
 4 |     println!("Hello, world!");
   |     ^^^^^^^
 ```
 
-发生此错误的原因是[`println`宏]是标准库的一部分，我们不再包含这个库。 因此我们无法再打印东西。这是有道理的，因为`println`写入[标准输出][standard output]，这是操作系统提供的特殊文件描述符。
-
-[`println` macro]: https://doc.rust-lang.org/std/macro.println.html
-[standard output]: https://en.wikipedia.org/wiki/Standard_streams#Standard_output_.28stdout.29
+出现这个错误的原因是，[println!宏](https://doc.rust-lang.org/std/macro.println.html)是标准库的一部分，而我们的项目不再依赖于标准库。我们选择不再打印字符串。这也能解释得通，因为`println!`将会向**标准输出**（[standard output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_.28stdout.29)）打印字符，它依赖于特殊的文件描述符，而这是由操作系统提供的特性。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
出现这个错误的原因是：[println!宏](https://doc.rust-lang.org/std/macro.println.html)是标准库的一部分，而我们的项目不再依赖于标准库。我们选择不再打印字符串。这也能解释得通，因为`println!`将会向**标准输出**（[standard output](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_.28stdout.29)）打印字符，它依赖于特殊的文件描述符，而这是由操作系统提供的特性。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 6]
```diff
@@ -111,50 +93,37 @@ error: `#[panic_handler]` function required, but not found
...
-/// This function is called on panic.
+/// 这个函数将在panic时被调用
 #[panic_handler]
 fn panic(_info: &PanicInfo) -> ! {
     loop {}
 }
 ```
 
-[`PanicInfo`参数][PanicInfo]包含发生异常的文件和行以及可选的异常消息。该函数永远不应该返回，因此通过返回[“never” type] `!`将其标记为[diverging function]。 目前，我们无法在此函数中执行太多操作，因此我们只是做无限循环。
+类型为[PanicInfo](https://doc.rust-lang.org/nightly/core/panic/struct.PanicInfo.html)的参数包含了panic发生的文件名、代码行数和可选的错误信息。这个函数从不返回，所以他被标记为**发散函数**（[diverging function](https://doc.rust-lang.org/book/first-edition/functions.html#diverging-functions)）。发散函数的返回类型称作**Never类型**（["never" type](https://doc.rust-lang.org/nightly/std/primitive.never.html)），记为`!`。对这个函数，我们目前能做的事情很少，所以我们只需编写一个无限循环`loop {}`。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
类型为[PanicInfo](https://doc.rust-lang.org/nightly/core/panic/struct.PanicInfo.html)的参数包含了panic发生的文件名、代码行数和可选的错误信息。这个函数从不返回，所以他被标记为**发散函数**（[diverging function](https://doc.rust-lang.org/book/first-edition/functions.html#diverging-functions)）。发散函数的返回类型称作**Never类型**（["never" type](https://doc.rust-lang.org/nightly/std/primitive.never.html)），记为`!`。对这个函数，我们目前能做的很少，所以我们只需编写一个无限循环`loop {}`。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 7]
```diff
@@ -111,50 +93,37 @@ error: `#[panic_handler]` function required, but not found
...
-[PanicInfo]: https://doc.rust-lang.org/nightly/core/panic/struct.PanicInfo.html
-[diverging function]: https://doc.rust-lang.org/1.30.0/book/first-edition/functions.html#diverging-functions
-[“never” type]: https://doc.rust-lang.org/nightly/std/primitive.never.html
+## eh_personality语言项
 
-## `eh_personality` 语言项
+语言项是一些编译器需求的特殊函数或类型。举例来说，Rust的[Copy](https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html) trait是一个这样的语言项，告诉编译器哪些类型需要遵循**复制语义**（[copy semantics](https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html)）——当我们查找`Copy` trait的[实现](https://github.com/rust-lang/rust/blob/485397e49a02a3b7ff77c17e4a3f16c653925cb3/src/libcore/marker.rs#L296-L299)时，我们会发现，一个特殊的`#[lang = "copy"]`属性将它定义为了一个语言项，达到与编译器联系的目的。
 
-语言项是编译器内部所需的特殊功能和类型。例如，[`Copy`]特征是一种语言项目，它告诉编译器哪些类型具有 [_copy语义_][`Copy`]。当我们查看[实现][copy code]时，我们看到它具有特殊的`#[lang = "copy"]`属性，将该属性定义为语言项。
+我们可以自己实现语言项，但这只应该是最后的手段：目前来看，语言项是高度不稳定的语言细节实现，它们不会经过编译期类型检查（所以编译器甚至不确保它们的参数类型是否正确）。幸运的是，我们有更稳定的方式，来修复上面的语言项错误。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
我们可以自己实现语言项，但这是下下策：目前来看，语言项是高度不稳定的语言细节实现，它们不会经过编译期类型检查（所以编译器甚至不确保它们的参数类型是否正确）。幸运的是，我们有更稳定的方式，来修复上面的语言项错误。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 8]
```diff
@@ -111,50 +93,37 @@ error: `#[panic_handler]` function required, but not found
...
 
-[`Copy`]: https://doc.rust-lang.org/nightly/core/marker/trait.Copy.html
-[copy code]: https://github.com/rust-lang/rust/blob/485397e49a02a3b7ff77c17e4a3f16c653925cb3/src/libcore/marker.rs#L296-L299
+`eh_personality`语言项标记的函数，将被用于实现**栈展开**（[stack unwinding](http://www.bogotobogo.com/cplusplus/stackunwinding.php)）。在使用标准库的情况下，当panic发生时，Rust将使用栈展开，来运行在栈上活跃的所有变量的**析构函数**（destructor）——这确保了所有使用的内存都被释放，允许调用程序的**父进程**（parent thread）捕获panic，处理并继续运行。但是，栈展开是一个复杂的过程，如Linux的[libunwind](http://www.nongnu.org/libunwind/)或Windows的**结构化异常处理**（[structured exception handling, SEH](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680657(v=vs.85).aspx)），通常需要依赖于操作系统的库；所以我们不在自己编写的操作系统中使用它。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
`eh_personality`语言项标记的函数，将被用于实现**栈展开**（[stack unwinding](http://www.bogotobogo.com/cplusplus/stackunwinding.php)）。在使用标准库的情况下，当panic发生时，Rust将使用栈展开，来运行在栈上活跃的所有变量的**析构函数**（destructor）——这确保了所有使用的内存都被释放，允许调用程序的**父进程**（parent thread）捕获panic，处理问题并继续运行。但是，栈展开是一个复杂的过程，如Linux的[libunwind](http://www.nongnu.org/libunwind/)或Windows的**结构化异常处理**（[structured exception handling, SEH](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680657(v=vs.85).aspx)），通常需要依赖于操作系统的库；所以我们不在自己编写的操作系统中使用它。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 9]
```diff
@@ -164,48 +133,43 @@ panic = "abort"
 panic = "abort"
 ```
 
-这会将`dev`配置文件（用于`cargo build`）和`release`配置文件（用于`cargo build --release`）的应急策略设置为`abort`。 现在，不再需要`eh_personality`语言项目了。
-
-[abort on panic]: https://github.com/rust-lang/rust/pull/32900
+这些选项能将**dev配置**（dev profile）和**release配置**（release profile）的panic策略设为`abort`。`dev`配置适用于`cargo build`，而`release`配置适用于`cargo build --release`。现在编译器应该不再要求我们提供`eh_personality`语言项实现。
 
-现在，我们修复了以上两个错误。 但是，如果我们现在尝试对其进行编译，则会发生另一个错误：
+现在我们已经修复了出现的两个错误，可以信心满满地开始编译了。然而，尝试编译运行后，一个新的错误出现了：
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
现在我们已经修复了出现的两个错误，可以尝试开始编译了。然而，尝试编译运行后，一个新的错误出现了：
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md [hunk 10]
```diff
@@ -251,265 +208,85 @@ release: 1.35.0-nightly
...
-          for architecture x86_64
-          clang: error: linker command failed with exit code 1 […]
-```
+我们将使用这个方法编写自己的操作系统内核。我们不将编译到`thumbv7em-none-eabihf`，而是使用描述`x86_64`环境的**自定义目标**（[custom target](https://doc.rust-lang.org/rustc/targets/custom.html)）。在下篇文章中，我们将详细描述一些相关的细节。
 
-macOS [does not officially support statically linked binaries] and requires programs to link the `libSystem` library by default. To override this and link a static binary, we pass the `-static` flag to the linker:
+### 链接器参数
 
-[does not officially support statically linked binaries]: https://developer.apple.com/library/content/qa/qa1118/_index.html
+我们也可以选择不编译到裸机系统，因为传递特定的参数也能解决链接器错误问题。虽然我们不将在后文中使用这个方法，为了教程的完整性，我们也撰写了专门的短文，来提供这个途径的解决方案。
```
Comment (Rustin-Liu) [inline blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]:
```suggestion
我们也可以选择不编译到裸机系统，因为传递特定的参数也能解决链接器错误问题。虽然我们不会在后文中使用这个方法，为了教程的完整性，我们也撰写了专门的短文，来提供这个途径的解决方案。
```

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md (hunk 2 above)
Comment (luojia65) [reply blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]: 可以可以

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md (hunk 7 above)
Comment (luojia65) [reply blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]: 这个可以

**Comment by Rustin-Liu** [comment]
I will fix some problems and review other file in this week.

**Comment by phil-opp** [comment]
@Rustin-Liu Thanks for reviewing! I don't speak Chinese, so it's difficult for me to evaluate your suggestions. 

@luojia65 Maybe you could give :+1: / :-1: reactions on the individual suggestions if you agree/disagree with them?

File: blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md (hunk 6 above)
Comment (luojia65) [reply blog/content/second-edition/posts/01-freestanding-rust-binary/index.zh-CN.md]: 应该都可以，在意译的范围内

**Comment by Rustin-Liu** [comment]
It seem I have no permissions push commit to this branch. So I need create a pr to rustcc-translation?

**Comment by phil-opp** [comment]
@Rustin-Liu It seems so, unfortunately.

**Comment by phil-opp** [comment]
Of course, Github suggestions work too. I just applied your existing suggestions.

**Comment by Rustin-Liu** [comment]
> Of course, Github suggestions work too. I just applied your existing suggestions.

pr:#700

**Comment by phil-opp** [comment]
Given that this pull request does not contain the CC BY-NC license yet, I have to ask you explicitly:

@Rustin-Liu @luojia65 

Are you willing to license your contributions under the CC BY-NC license too? This would be required for merging your translation included in this PR. If yes, please reply with:

> I hereby license my contributions to this pull request under the CC BY-NC license (https://creativecommons.org/licenses/by-nc/4.0/).

After you agreed, I will be able to merge the license file into this PR so that you won't need to explicitly agree again on future contributions.

Sorry for the hassle!

**Comment by luojia65** [comment]
I hereby license my contributions to this pull request under the CC BY-NC license (https://creativecommons.org/licenses/by-nc/4.0/).

**Comment by Rustin-Liu** [comment]
I hereby license my contributions to this pull request under the CC BY-NC license (https://creativecommons.org/licenses/by-nc/4.0/).

**Comment by phil-opp** [comment]
Thanks! I just merged the master branch, which means that additional contributions to this pull request are automatically licensed as CC BY-NC too.

**Comment by phil-opp** [comment]
Given that Zola 0.10 was released, I'm going to merge this pull request now so that we can publish it. Thanks for all your work on this!

@Rustin-Liu In case you want to refactor the 4th post too, feel free to open a pull request against master.

**Pull request merged**